buildSrc/plugins/src/main/kotlin/org/jetbrains/kotlin/TestDirectives.kt

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package org.jetbrains.kotlin
+
+import java.nio.file.Paths
+import java.util.regex.Pattern
+
+/**
+ * Creates files from the given source file that may contain different test directives.
+ *
+ * @return list of file names to be compiled
+ */
+fun KonanTest.buildCompileList(outputDirectory: String): List<String> {
+    val result = mutableListOf<String>()
+    val srcFile = project.file(source)
+    // Remove diagnostic parameters in external tests.
+    val srcText = srcFile.readText().replace(Regex("<!.*?!>(.*?)<!>")) { match -> match.groupValues[1] }
+
+    if (srcText.contains("// WITH_COROUTINES")) {
+        val coroutineHelpersFileName = "$outputDirectory/helpers.kt"
+        createFile(coroutineHelpersFileName, createTextForHelpers(true))
+        result.add(coroutineHelpersFileName)
+    }
+
+    val filePattern = Pattern.compile("(?m)// *FILE: *(.*)")
+    val matcher = filePattern.matcher(srcText)
+
+    if (!matcher.find()) {
+        // There is only one file in the input
+        val filePath = "$outputDirectory/${srcFile.name}"
+        registerKtFile(result, filePath, srcText)
+    } else {
+        // There are several files
+        var processedChars = 0
+        while (true) {
```

Comment:
It looks like we can replace this loop with a `do..while` one:
```kotlin
do {
    ...
    val nextFileExists = matcher.find()
    ...
} while(!nextFileExists)
```

Reply:
It actually can be rewritten using stdlib functions for parsing. This is a straight forward conversion and I'm going to return to it later with rewriting external tests.